File: web/src/domain/builders/EvaluacionBuilder.ts
import { Evaluacion, EvaluacionSchema } from "../models/Evaluacion";
import { v4 as uuidv4 } from "uuid";

export class EvaluacionBuilder {
  private _evaluacion: Partial<Evaluacion> = {
    id: uuidv4(),
    fecha: new Date().toISOString().split("T")[0],
  };

  withPaciente(id: string) {
    this._evaluacion.paciente_id = id;
    return this;
  }

  withOptica(id: string) {
    this._evaluacion.optica_id = id;
    return this;
  }

  withRefraccionOd(esf: string, cil: string, eje: string) {
    const normEsf = this.normalizeSphere(esf);
    const { nEsf, nCil, nEje } = this.transpose(normEsf, cil, eje);
    this._evaluacion.receta_od_esf = nEsf;
    this._evaluacion.receta_od_cil = nCil;
    this._evaluacion.receta_od_eje = nEje;
    this.updateAutoDiagnosis();
    return this;
  }

  withRefraccionOi(esf: string, cil: string, eje: string) {
    const normEsf = this.normalizeSphere(esf);
    const { nEsf, nCil, nEje } = this.transpose(normEsf, cil, eje);
    this._evaluacion.receta_oi_esf = nEsf;
    this._evaluacion.receta_oi_cil = nCil;
    this._evaluacion.receta_oi_eje = nEje;
    this.updateAutoDiagnosis();
    return this;
  }

  build(): Evaluacion {
    return EvaluacionSchema.parse(this._evaluacion);
  }

  private normalizeSphere(value: string): string {
    const v = value.trim().toLowerCase();
    return (v === "plano" || v === "neutro" || v === "pl") ? "0.00" : value;
  }

  private updateAutoDiagnosis() {
    // Aquí iría la lógica de presbicia y anisometropía similar a Android
    const eeOd = this.calculateEE(this._evaluacion.receta_od_esf || "0", this._evaluacion.receta_od_cil || "0");
    const eeOi = this.calculateEE(this._evaluacion.receta_oi_esf || "0", this._evaluacion.receta_oi_cil || "0");
    const diff = Math.abs(eeOd - eeOi);
    // Para simplificar, asumimos campos en el esquema (que deberíamos expandir)
  }

  private calculateEE(esf: string, cil: string): number {
    const e = parseFloat(esf) || 0;
    const c = parseFloat(cil) || 0;
    return e + (c / 2.0);
  }

  private transpose(esf: string, cil: string, eje: string) {
    const c = parseFloat(cil) || 0;
    if (c <= 0) return { nEsf: esf, nCil: cil, nEje: eje };

    const e = parseFloat(esf) || 0;
    const a = parseInt(eje) || 0;

    const nEsf = (e + c).toFixed(2);
    const nCil = (-c).toFixed(2);
    let nEje = a + 90;
    if (nEje > 180) nEje -= 180;

    return { nEsf, nCil, nEje: nEje.toString() };
  }
}

File: app/src/main/java/com/example/optoapp/domain/builder/EvaluacionBuilder.kt
package com.example.optoapp.domain.builder

import com.example.optoapp.domain.model.EvaluacionModel
import java.time.LocalDate
import java.util.UUID

class EvaluacionBuilder {
    private var evaluacion = EvaluacionModel(
        id = UUID.randomUUID().toString(),
        pacienteId = "",
        fecha = LocalDate.now(),
        opticaId = ""
    )

    fun id(id: String) = apply { evaluacion = evaluacion.copy(id = id) }
    fun pacienteId(id: String) = apply { evaluacion = evaluacion.copy(pacienteId = id) }
    fun fecha(fecha: LocalDate) = apply { evaluacion = evaluacion.copy(fecha = fecha) }
    fun opticaId(id: String) = apply { evaluacion = evaluacion.copy(opticaId = id) }

    fun refraccionOd(esf: String, cil: String, eje: String) = apply {
        val normalizedEsf = normalizeSphere(esf)
        val (nEsf, nCil, nEje) = transposeIfNeeded(normalizedEsf, cil, eje)
        evaluacion = evaluacion.copy(
            recetaOdEsf = nEsf,
            recetaOdCil = nCil,
            recetaOdEje = nEje
        )
        updateAutoDiagnosis()
    }

    fun refraccionOi(esf: String, cil: String, eje: String) = apply {
        val normalizedEsf = normalizeSphere(esf)
        val (nEsf, nCil, nEje) = transposeIfNeeded(normalizedEsf, cil, eje)
        evaluacion = evaluacion.copy(
            recetaOiEsf = nEsf,
            recetaOiCil = nCil,
            recetaOiEje = nEje
        )
        updateAutoDiagnosis()
    }

    fun add(addOd: String, addOi: String) = apply {
        evaluacion = evaluacion.copy(
            addCercaOd = addOd,
            addCercaOi = addOi
        )
        updateAutoDiagnosis()
    }

    fun agudezaVisual(od: String, oi: String) = apply {
        evaluacion = evaluacion.copy(
            recetaOdAv = od,
            recetaOiAv = oi
        )
        updateAutoDiagnosis()
    }

    fun build(): EvaluacionModel {
        validate()
        return evaluacion
    }

    private fun normalizeSphere(value: String): String {
        val v = value.trim().lowercase()
        return if (v == "plano" || v == "neutro" || v == "pl") "0.00" else value
    }

    private fun updateAutoDiagnosis() {
        // Solo automatizar si no están en modo manual
        if (evaluacion.autoPresbicia) {
            val hasAdd = evaluacion.addCercaOd.toDoubleOrNull() ?: 0.0 > 0.0 || 
                         evaluacion.addCercaOi.toDoubleOrNull() ?: 0.0 > 0.0
            evaluacion = evaluacion.copy(otrosPresbicia = hasAdd)
        }

        if (evaluacion.autoAnisometropia) {
            val eeOd = calculateEE(evaluacion.recetaOdEsf, evaluacion.recetaOdCil)
            val eeOi = calculateEE(evaluacion.recetaOiEsf, evaluacion.recetaOiCil)
            val diff = kotlin.math.abs(eeOd - eeOi)
            evaluacion = evaluacion.copy(otrosAnisometropia = diff >= 2.00)
        }

        // Ambliopía: Simplificado para el ejemplo (diferencia de líneas)
        // En una implementación real se usaría un mapa de valores LogMAR o líneas Snellen
    }

    private fun calculateEE(esf: String, cil: String): Double {
        val e = esf.toDoubleOrNull() ?: 0.0
        val c = cil.toDoubleOrNull() ?: 0.0
        return e + (c / 2.0)
    }

    private fun validate() {
        if (evaluacion.pacienteId.isBlank()) throw IllegalStateException("Paciente ID es obligatorio")
        if (evaluacion.opticaId.isBlank()) throw IllegalStateException("Optica ID es obligatorio")
    }

    /**
     * Lógica de transposición: convierte cilindro positivo a negativo.
     * Nueva Esfera = Esfera + Cilindro
     * Nuevo Cilindro = -Cilindro
     * Nuevo Eje = Eje + 90 (si > 180, restar 180)
     */
    private fun transposeIfNeeded(esf: String, cil: String, eje: String): Triple<String, String, String> {
        val c = cil.toDoubleOrNull() ?: 0.0
        if (c <= 0) return Triple(esf, cil, eje)

        val e = esf.toDoubleOrNull() ?: 0.0
        val a = eje.toIntOrNull() ?: 0

        val newEsf = e + c
        val newCil = -c
        var newEje = a + 90
        if (newEje > 180) newEje -= 180

        return Triple(
            String.format("%.2f", newEsf),
            String.format("%.2f", newCil),
            newEje.toString()
        )
    }
}
